src/rustllvm/PassWrapper.cpp

```diff
@@ -503,8 +504,133 @@ LLVMRustWriteOutputFile(LLVMTargetMachineRef Target, LLVMPassManagerRef PMR,
   return LLVMRustResult::Success;
 }
 
+
+// Callback to demangle function name
+// Parameters:
+// * name to be demangled
+// * name len
+// * output buffer
+// * output buffer len
+// Returns len of demangled string, or 0 if demangle failed.
+typedef size_t (*DemangleFn)(const char*, size_t, char*, size_t);
+
+
+namespace {
+
+class RustAssemblyAnnotationWriter : public AssemblyAnnotationWriter {
+  DemangleFn Demangle;
+  static const size_t BufCap = 500;
+  char Buf[BufCap];
+
+public:
+  RustAssemblyAnnotationWriter(DemangleFn Demangle) : Demangle(Demangle) {}
+
+  // return empty string if demangle failed
+  // or if name does not need to be demangled
+  StringRef CallDemangle(StringRef name) {
+    if (!Demangle) {
+      // not possible, but it is safer
+      return StringRef();
+    }
+
+    auto R = Demangle(name.data(), name.size(), Buf, BufCap);
+    if (!R) {
+      // demangle failed
+      return StringRef();
+    }
+
+    auto Demangled = StringRef(Buf, R);
+    if (Demangled == name) {
+      return StringRef();
+    }
+
+    return Demangled;
+  }
+
+  void emitFunctionAnnot(const Function *F,
+                         formatted_raw_ostream &OS) override {
+    StringRef Demangled = CallDemangle(F->getName());
+    if (Demangled.empty()) {
+        return;
+    }
+
+    OS << "; " << Demangled << "\n";
+  }
+
+  void emitInstructionAnnot(const Instruction *I,
+                            formatted_raw_ostream &OS) override {
+    const char *Name;
+    const Value *Value;
+    if (const CallInst *CI = dyn_cast<CallInst>(I)) {
```

Comment:
Some other cases where a demangle could be relevant is taking the address of the function.

That would be `store %place, @function`.

Maybe it is worth to always look at the instruction operands and output a demangled version if there’s a function operand?